R/bs-dependencies.R

```diff
@@ -124,39 +124,133 @@ bs_dependencies <- function(
   )
 }
 
-#' @rdname bs_dependencies
-#' @param func a function that takes a [bs_theme()] object (or `theme_default`) as input and
-#' and returns an [htmlDependency()] (or a list of them).
-#' @param theme_default In the case that `func` wants to be executed outside of a Shiny
-#' runtime (e.g., in an R Markdown doc), an expression that yields a value to provide to `func`.
-#' @param env environment for evaluation the expression provided to `theme_default`.
+
+#' Create a Bootstrap dependency
+#'
 #' @export
-bs_runtime_dependencies <- function(func, theme_default = bs_global_get(), envir = parent.env()) {
-  theme_default <- substitute(theme_default)
+bs_dependency <- function(input = list(), theme, name, version,
+  cache_key_extra = NULL, .dep_args = list(), .sass_args = list())
+{
+  sass_args <- c(
+    list(
+      rules = input,
+      layer = theme,
+      output = sass::output_template(basename = name, dirname = name),
+      write_attachments = TRUE,
+      cache_key_extra = cache_key_extra
+    ),
+    .sass_args
+  )
+  outfile <- do.call(sass::sass_partial, sass_args)
+
+  dep_args <- c(
+    list(
+      name = name,
+      version = version,
+      src = dirname(outfile),
+      stylesheet = basename(outfile)
+    ),
+    .dep_args
+  )
+  do.call(htmlDependency, dep_args)
+}
+
+
+#' Note that `func` should not be created in a closure -- for Shiny to
+#' deduplicate the html dependencies, it needs to be able to tell that the
+#' `func` from a previous invocation is `identical()` to the `func` in later
+#' invocations. In order to that, the
+#'
+#' @param func a function that takes a [bs_theme()] object (or `theme_default`)
+#'   as input and and returns an [htmlDependency()] (or a list of them).
+#' @export
+#'
+#' @examples
+#'
+#'
+#' myWidgetVersion <- "1.2.3"
+#'
+#' myWidgetDependency <- function() {
+#'   list(
+#'     bs_dependency_dynamic(myWidgetCss),
+#'     htmlDependency(
+#'       name = "mywidget-js",
+#'       version = myWidgetVersion,
+#'       src = system.file(package = "mypackage", "js"),
+#'       script = "mywidget.js"
+#'     )
+#'   )
+#' }
+#'
+#' myWidgetCSS <- function(theme) {
+#'   if (!is_bs_theme(theme)) {
+#'     return(
+#'       htmlDependency(
+#'         name = "mywidget-css",
+#'         version = myWidgetVersion,
+#'         src = system.file(package = "mypackage", "css"),
+#'         stylesheet = "mywidget.css"
+#'       )
+#'     )
+#'   }
```

Comment:
I was thinking it might make sense for `bs_dependency_dynamic()` to only ever call `func` if `theme` is a `bs_theme()`. I don't feel that strongly about it, but it seems like it might be more intuitive to keep your static vs dynamic dependency logic separated?

Reply:
Ohh, thinking about it more, I suppose there could be a situation where an update to a `NULL` theme should trigger a restyle, so I think this is the right approach

Reply:
I was initially headed in the direction of always including the static CSS, and having this part either return `NULL` (if `theme` is not a `bs_theme`) or the compiled Sass, with the idea that the compiled Sass would override the static CSS, but there were two problems:

* The `version` for the compiled Sass wouldn't cause it to override the static CSS
* The code that that authors would write was really convoluted. I can provide an example a bit later.